File: 2024/day 25/AOC25.py
def parse_schematics(file_path):
    """Parse the input file to extract lock and key schematics."""
    with open(file_path, 'r') as f:
        data = f.read().strip().split("\n\n")

    locks = []
    keys = []

    for schematic in data:
        lines = schematic.split("\n")
        if lines[0] == "#####":  # Identify locks by the top row filled with '#'
            locks.append(lines)
        elif lines[-1] == "#####":  # Identify keys by the bottom row filled with '#'
            keys.append(lines)

    return locks, keys

def convert_to_heights(schematic, is_lock=True):
    """Convert a schematic to a list of heights."""
    num_columns = len(schematic[0])
    heights = []

    for col in range(num_columns):
        height = 0
        if is_lock:
            for row in range(len(schematic)):
                if schematic[row][col] == '#':
                    height += 1
                else:
                    break
        else:  # For keys
            for row in range(len(schematic) - 1, -1, -1):
                if schematic[row][col] == '#':
                    height += 1
                else:
                    break
        heights.append(height)

    return heights

def count_valid_pairs(locks, keys):
    """Count the number of valid lock-key pairs."""
    valid_pairs = 0
    max_height = len(locks[0])  # Total rows in the schematics

    for lock in locks:
        lock_heights = convert_to_heights(lock, is_lock=True)
        for key in keys:
            key_heights = convert_to_heights(key, is_lock=False)

            # Check if all columns fit
            fits = all(lh + kh <= max_height for lh, kh in zip(lock_heights, key_heights))
            if fits:
                valid_pairs += 1

    return valid_pairs

def main():
    file_path = "christmas.txt"  
    locks, keys = parse_schematics(file_path)
    result = count_valid_pairs(locks, keys)
    print(f"Number of unique lock/key pairs that fit: {result}")

if __name__ == "__main__":
    main()

File: 2024/day 25/AOC25.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func parseSchematics(filePath string) ([][]string, [][]string) {
	file, err := os.Open(filePath)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return nil, nil
	}
	defer file.Close()

	content := ""
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		content += scanner.Text() + "\n"
	}

	// Handle both Unix and Windows line endings
	contentStr := strings.ReplaceAll(content, "\r\n", "\n")
	schematics := strings.Split(strings.TrimSpace(contentStr), "\n\n")
	var locks, keys [][]string

	for _, schematic := range schematics {
		lines := strings.Split(schematic, "\n")
		if len(lines) > 0 {
			if lines[0] == "#####" {
				locks = append(locks, lines)
			} else if lines[len(lines)-1] == "#####" {
				keys = append(keys, lines)
			}
		}
	}

	return locks, keys
}

func convertToHeights(schematic []string, isLock bool) []int {
	if len(schematic) == 0 {
		return nil
	}

	numColumns := len(schematic[0])
	heights := make([]int, numColumns)

	for col := 0; col < numColumns; col++ {
		height := 0
		if isLock {
			for row := 0; row < len(schematic); row++ {
				if schematic[row][col] == '#' {
					height++
				} else {
					break
				}
			}
		} else {
			for row := len(schematic) - 1; row >= 0; row-- {
				if schematic[row][col] == '#' {
					height++
				} else {
					break
				}
			}
		}
		heights[col] = height
	}

	return heights
}

func countValidPairs(locks, keys [][]string) int {
	if len(locks) == 0 {
		return 0
	}

	validPairs := 0
	maxHeight := len(locks[0])

	for _, lock := range locks {
		lockHeights := convertToHeights(lock, true)
		for _, key := range keys {
			keyHeights := convertToHeights(key, false)

			fits := true
			for i := range lockHeights {
				if lockHeights[i]+keyHeights[i] > maxHeight {
					fits = false
					break
				}
			}

			if fits {
				validPairs++
			}
		}
	}

	return validPairs
}

func main() {
	filePath := "christmas.txt"
	locks, keys := parseSchematics(filePath)
	result := countValidPairs(locks, keys)
	fmt.Printf("Number of unique lock/key pairs that fit: %d\n", result)
}
